### ISSUE TITLE
[BUG]A Use After Free Bug in ssl/ssl_lib.c

### ISSUE THREAD

**Issue by Kylin-1** [openssl/openssl#20299]
**File:**  ssl/ssl_lib.c
**Bug Function:**  void SSL_free
**Version:** Git-master(2023-02-11)

**Description:** According to the structure definition, a structure `SSL` is included in the structure `SSL_CONNECTION`.
In the function `SSL *ossl_ssl_connection_new_int(...)` we define two Pointers  `SSL CONNECTION *s` and `SSL *ssl`.
At line 735,`ssl = &s->ssl` makes  `ssl` point to `s->ssl`.
At line 737, `s` was freed. But the pointer variable `ssl` still points to a block of space address of `s`.
In function `SSL_free(SSL *s)`, line 1327, `CRYPTO_DOWN_REF(&s->references, &i, s->lock)` visited the elements of `s`.
Similar to `ssl` on line 735, if this variable `s` points to a structure `SSL` which was insided in a sturct `SSL_CONNECTION`, this will cause a **use after free bug** when function `SSL *ossl_ssl_connection_new_int(...)` was run before function `SSL_free(SSL *s)`.

**Code Description:**
```
SSL *ossl_ssl_connection_new_int(……)
{
728:    SSL_CONNECTION *s;
729:    SSL *ssl; 
    ……
735:    ssl = &s->ssl;     //ssl points to s->ssl.
    if (!ossl_ssl_init(ssl, ctx, method, SSL_TYPE_SSL_CONNECTION)) {
737:        OPENSSL_free(s);    //s was freed here!
        s = NULL;
        goto sslerr;
    }
    ……
 err:
    SSL_free(ssl);    //Attention: although s was freed, ssl still be visited here.
    return NULL;
}

……

void SSL_free(SSL *s)
{
    int i;

    if (s == NULL)
        return;
1327：    CRYPTO_DOWN_REF(&s->references, &i, s->lock);    //This may cause a use after free bug!
    ……
}

**Comment by blazmonster** [comment]
Hey,
I will work on the fix.

**Comment by tmshort** [comment]
Fixed by #20300
